[HxMultiSelect] filtering and select all

bholbrook opened the pull request:
# Summary of changes
- Added filtering capability to the HxMultiSelect component as per [Issue 616](https://github.com/havit/Havit.Blazor/issues/616)

- Added select all capability to the HxMultiSelect component as per [Issue 453](https://github.com/havit/Havit.Blazor/issues/453)

Clicking when unchecked will select all items in the currently filtered list.
Clicking when checked will deselect all items in the currently filtered list.
Clicking a checked item or entering any text into the filter will uncheck the select all.
Manually clicking and selecting all items in the filtered list will not check the select all.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
Please do not change the order of members in the class. Our coding convention is:
1. parameters (in order of importance)
2. `[Inject]` - dependencies
3. private fields
4. constructor + other methods (usually in order of their appearance in component lifecycle)

bholbrook replied:
Styling has been reverted back to the original.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/HxMultiSelect.cs:
```diff
@@ -110,6 +110,11 @@ public class HxMultiSelect<TValue, TItem> : HxInputBase<List<TValue>>, IInputWit
 	/// </summary>
 	[Parameter] public RenderFragment InputGroupEndTemplate { get; set; }
 
+	[Parameter] public bool EnableFiltering { get; set; }
```
We don't have any other `EnableXy` parameters for optional features.
This should probably be `Filterable="true"` or `AllowFiltering="true"` or `Filtering="true"`. Not sure now. :-D
(We can change this later before merging to master.)

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/HxMultiSelect.cs:
```diff
@@ -202,7 +221,14 @@ private string GetInputText()
-		else if ((InputTextSelector is null) || (Data is null) || (CurrentValue is null))
+
+		// If filtering is enabled and the dropdown is visible then we want to display the user-entered filter text instead of the selected values summary
+		if (EnableFiltering && IsShown)
+		{
+			return InputText;
+		}
```
(Oops, we misunderstood each other regarding the implementation. I will explain in general comment.)

bholbrook replied:
Thank you for the comments and feedback. I would still like to implement these changes and while I believe the way I've implemented the omni-box to be suitable for my scenario I understand it may not be suitable for everyone.
I shall rework this pull request within the next day or so to meet your standards and suggestions.

bholbrook replied:
I have reworked this component and believe it's in a good place to re-review.

bholbrook replied:
I've extended this to also include the select all functionality as requested in [Issue 453](https://github.com/havit/Havit.Blazor/issues/453).

hakenr reviewed (changes requested):
I'm sorry, but I misunderstood the intent, we kind of missed each other. I thought we were talking about a new search field inside the dropdown and I didn't notice that you were describing using an existing input for filtering.
We would actually like to teach the component this UX:
![image](https://github.com/havit/Havit.Blazor/assets/12828888/d7c089ff-f060-4ec4-8efa-d6a27bd80ae6)
![image](https://github.com/havit/Havit.Blazor/assets/12828888/d004f8a2-08ab-4e2b-8561-f84871ba1bf4)

In our opinion, the filtering functionality combined with the main input is confusing because the selected items are not visible when filtering.
![image](https://github.com/havit/Havit.Blazor/assets/12828888/68983de0-9c30-433f-bc6a-f87015860242)

I am very sorry, but we cannot accept the filtering functionality in this form. If you would still be willing to give your time to rework it into the intended form, we would be very grateful. Otherwise, we will try to move the request higher in the backlog to appreciate the energy put into the functionality and get the filtering into the component sooner.

bholbrook commented on BlazorAppTest/Pages/HxMultiSelectTest.razor:
```diff
@@ -1,38 +1,61 @@
+
+    private bool FilterSelector(CultureInfo item, string filter)
+    {
+        return string.IsNullOrWhiteSpace(filter) || item.EnglishName.Contains(filter, StringComparison.OrdinalIgnoreCase);
```
I've changed this and the new default FilterPredicate to use string.IsNullOrEmpty.

bholbrook commented on Havit.Blazor.Components.Web.Bootstrap/Forms/HxMultiSelect.cs:
```diff
@@ -110,6 +110,11 @@ public class HxMultiSelect<TValue, TItem> : HxInputBase<List<TValue>>, IInputWit
+
+	[Parameter] public Func<TItem, string, bool> FilterSelector { get; set; } = (_, _) => true;
+
+	private bool IsShown { get; set; }
```
A property isn't required. This has been changed to a field.

hakenr replied:
Please rename to `isDropdownShown` to better refer to the dropdown part of the component.

hakenr commented:
@bholbrook Is the PR ready for review now or are you still working on the code?

bholbrook commented:
@hakenr It's ready to be re-reviewed when you can please :)

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor:
```diff
@@ -36,25 +34,60 @@
+                            <div class="form-check">
+                                <input class="form-check-input" type="checkbox" id="@($"{InputId}_selectall")" checked="@selectAll" tabindex="-1" />
+                                <label class="form-check-label" for="@($"{InputId}_selectall")" @onclick:preventDefault>
+                                    @(SelectAllText ?? "-- select all --")
```
We have to provide localization for the default text "select all" from the `IStringLocalizer<>`.

bholbrook replied:
I've reworked this to call a helper method which uses the newly injected IStringLocalizer with new resource file. I've only added English for now as that's all I'm comfortable with.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -14,6 +16,8 @@ public partial class HxMultiSelectInternal<TValue, TItem>
 
 	[Parameter] public List<int> SelectedIndexes { get; set; }
```
Please remove the parameter, this is dead code (it was no longer being used before the PR).

bholbrook replied:
I have removed this property and the Values property which was also not in use.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
 	/// </summary>
 	[Parameter(CaptureUnmatchedValues = true)] public Dictionary<string, object> AdditionalAttributes { get; set; }
 
+	[Inject] private IJSRuntime JSRuntime { get; set; } = default!;
```
Remove nullable-references notation, we do not use nullable-references feature in our code.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+	/// <inheritdoc cref="ComponentBase.OnAfterRenderAsync(bool)" />
+	protected override async Task OnAfterRenderAsync(bool firstRender)
+	{
+		await base.OnAfterRenderAsync(firstRender);
```
No need to call empty `base` method. We are in a library code where we can be a little bit more aggressive.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+		jsModule ??= await JSRuntime.ImportHavitBlazorBootstrapModuleAsync(nameof(HxMultiSelect));
+	}
+
+	private async Task HandleItemSelectionChangedAsync(bool newChecked, TItem item, bool triggerSelectAllEvent = true)
 	{
 		await ItemSelectionChanged.InvokeAsync(new SelectionChangedArgs
 		{
 			Checked = newChecked,
 			Item = item
 		});
+
+		if (triggerSelectAllEvent)
+		{
+			await ChangeSelectAllAsync(false);
+		}
+	}
```
(I think we should not expose the `SelectAllChanged` event (`EventCallback`) as parameter et al.)
In case we keep the `SelectAllChanged`, please change the method argument name to `resetSelectAll` to better describe the meaning.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+		return SelectedValues?.Contains(value) ?? false;
+	}
+
+	private Task ChangeSelectAllAsync(bool selectAll)
+	{
+		this.selectAll = selectAll;
+		return SelectAllChanged.InvokeAsync(selectAll);
+	}
```
(I think we should not expose the SelectAllChanged event (EventCallback) as parameter et al.)

In case we keep the `SelectAllChanged` parameter, we should check if the `selectAll` value changed and invoke the `EventCallback` only if the value really changed. (Now it gets triggered from `HandleItemSelectionChangedAsync` with every single selection of individual item in the list.)

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+		}
+	}
+
+	private bool selectAll;
```
(Class members order.) Please move the field definition above all the methods where all the other fields are.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+
+	private bool selectAll;
+
+	private async Task HandleSelectAllClickedAsync()
+	{
+		var filteredItems = GetFilteredItems();
+
+		// If all items are already selected then they should be deselected, otherwise only records that aren't selected should be
+		if (selectAll)
+		{
+			foreach (var item in filteredItems)
+			{
+				// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+				await HandleItemSelectionChangedAsync(false, item, false);
+			}
+		}
+		else
+		{
+			foreach (var item in filteredItems)
+			{
+				var value = SelectorHelpers.GetValue<TItem, TValue>(ValueSelector, item);
+				var itemSelected = DoSelectedValuesContainValue(value);
+
+				// If the item is already selected we don't need to reselect it
+				if (!itemSelected)
+				{
+					// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+					await HandleItemSelectionChangedAsync(!itemSelected, item, false);
+				}
+			}
+		}
+
+		await ChangeSelectAllAsync(!selectAll);
+	}
```
Not sure we can change the `selectAll` value at the end of this method as it contains many `await` for asynchronnous calls and this might lead to rendering of the component with old `selectAll` value and some other unwanted effects.

Also see overall comment regarding the `ItemSelectionChanged` callback for select-all scenario.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+		return SelectAllChanged.InvokeAsync(selectAll);
+	}
+
+	private Task HandleInputChangedAsync(ChangeEventArgs e)
```
Please rename to `HandleFilterInputChangedAsync' to refer to the filter explicitly.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs:
```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+		await InvokeOnShownAsync(this.InputId);
+	}
+
+	/// <summary>
+	/// Collapses the item.
+	/// </summary>
+	public async Task HideAsync()
+	{
+		await EnsureJsModuleAsync();
+		await jsModule.InvokeVoidAsync("hide", elementReference);
+	}
+
+	/// <summary>
+	/// Expands the item.
+	/// </summary>
+	public async Task ShowAsync()
+	{
+		await EnsureJsModuleAsync();
+		await jsModule.InvokeVoidAsync("show", elementReference);
+	}
```
Please remove the methods. They are not being used nor exposed to the user.

bholbrook replied:
Have removed these methods.

hakenr commented on Havit.Blazor.Components.Web.Bootstrap/wwwroot/HxMultiSelect.js:
```diff
@@ -0,0 +1,50 @@
+﻿export function initialize(element, hxMultiSelectDotnetObjectReference, shouldToggle) {
+	if (!element) {
+		return;
+	}
+
+	element.hxMultiSelectDotnetObjectReference = hxMultiSelectDotnetObjectReference;
+	element.addEventListener('shown.bs.dropdown', handleDropdownShown);
+	element.addEventListener('hidden.bs.dropdown', handleDropdownHidden);
+
+	var d = new bootstrap.Dropdown(element, {
+		toggle: shouldToggle,
+	});
+}
```
Please remove the `shouldToggle` option, it is always `false`.

bholbrook replied:
Have removed the toggle option

hakenr reviewed (commented):
@bholbrook First off, I have to commend the effort you've put into working on changes of this scale. I myself was surprised by the sheer amount of changes, which is partly due to the combination with adding the "select all" feature. This raises a lot of new questions, although I understand that implementing both features together is more efficient.

For now, I've added some inline code comments and have noted the following points (for myself) that need to be addressed. Most of these will require further thought on my part, and some will be discussed with colleagues in our internal design meeting:

- [ ] `AllowFiltering`, `AllowSelectAll` naming?
- [x] `EmptyTemplate` naming. Perhaps `FilterEmptyResultTemplate` and `FilterEmptyResultText` (with defaults) for easier usage?
- [ ] SelectAll when filtered - should it select only visible items? Is that what we want?
- [ ] Filter-box - should it have a search icon and a clear icon?
- [x] `ClearFilterOnHide` - should the default be `true`?
- [x] `OnShown` and `OnHidden` - do we even need these parameters? How would anyone use them?
- [x] `MultiSelectSettings` - `AllowFiltering`, `AllowSelectAll`, `ClearFilterOnHide`, ?`FilterPredicate`?
- [ ] `SelectAllChanged` - do we even need such an `EventCallback`? How would it be used? If so, maybe `OnSelectAllChanged`, and it should not have a value.
- [ ] Missing documentation for new parameters
- [ ] Missing demo documentation for new features
- [ ] Default "-- select all --" text - do we want those dashes? If so, probably just one on each side?
- [ ] `ItemSelectionChanged` - we probably can't use this for the SelectAll scenario, as it would trigger a series of `ValueChanged` for each selected item, whereas it might be more appropriate to trigger `ValueChanged` just once with a `List<TValue>` of all items
- [ ] `InputSize` to be applied to inner filter-input

Looking forward to your thoughts on these points.

hakenr commented:
@crdo, this is a significant contribution that we shouldn't let sit idle. Could you please take a look at the PR (UX, HTML, API, ...) when you get a chance? Thanks!

bholbrook commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor:
```diff
@@ -36,25 +34,60 @@
+                <input @ref="filterInputReference"
+                       id="@($"{InputId}_filter")"
+                       type="text"
+                       class="@CssClassHelper.Combine("form-control")"
```
Have removed the redundant call.

bholbrook commented:
@hakenr In response to your list, these are my thoughts.

* [ ]  `AllowFiltering`, `AllowSelectAll` naming?
I don't mind what these are called as long as it's obvious they are a boolean on/off feature using an action verb prefix.
Happy to change them to whatever you want, but I'd rather have the functionality committed than a perfected name.

* [ ]  `EmptyTemplate` naming. Perhaps `FilterEmptyResultTemplate` and `FilterEmptyResultText` (with defaults) for easier usage?
Have renamed the existing template parameter to FilterEmptyResultTemplate.
Have added a new FilterEmptyResultText parameter.
Have added a default empty template utilising this new FilterEmptyResultText parameter. This also uses localisation.

* [ ]  SelectAll when filtered - should it select only visible items? Is that what we want?
I definitely think so. You have to actively filter and you have to have to actively click the select all button, which means when you do click select all you've actively chosen to select what you see contextually.
I can't think of a single real-world scenario when you'd want to select every item in the list while you've got a filter.

* [ ]  Filter-box - should it have a search icon and a clear icon?
A search icon isn't required as it will update the list immediately as the user types.
A clear icon has now been added and has been added to the settings too.
[IN_PROGRESS] Icon should be placed inside the filter input box.

* [ ]  `ClearFilterOnHide` - should the default be `true`?
I have changed the default to true.
As a note: When the dropdown is expanded the filter is automatically highlighted so the user can just hit delete so even if they do actively set it to false the filter is still very usable.

* [ ]  `OnShown` and `OnHidden` - do we even need these parameters? How would anyone use them?
I have removed these.

* [ ]  `MultiSelectSettings` - `AllowFiltering`, `AllowSelectAll`, `ClearFilterOnHide`, ?`FilterPredicate`?
Added AllowFiltering and AllowSelectAll to the MultiSelectSettings and set both defaults to false.
Added ClearFilterOnHide to the MultiSelectSettings and set the default to true.

* [ ]  `SelectAllChanged` - do we even need such an `EventCallback`? How would it be used? If so, maybe `OnSelectAllChanged`, and it should not have a value.
I have removed this.

* [ ]  Missing documentation for new parameters
All documentation has been added to `HxMultiSelect` and removed from `HxMultiSelectInternal`

* [ ]  Missing demo documentation for new features
I have added new examples for basic filtering, custom filtering, and a simple select all example.
I don't know how to run the documentation project locally so haven't been able to test these :(
How can I run the documentation to check please?

* [ ]  Default "-- select all --" text - do we want those dashes? If so, probably just one on each side?
I've changed this to default to "-select all-".

* [ ]  `ItemSelectionChanged` - we probably can't use this for the SelectAll scenario, as it would trigger a series of `ValueChanged` for each selected item, whereas it might be more appropriate to trigger `ValueChanged` just once with a `List<TValue>` of all items

hakenr commented:
@bholbrook

### SelectAll when filtered - should it select only visible items? Is that what we want?
>I definitely think so. You have to actively filter and you have to have to actively click the select all button, which means when you do click select all you've actively chosen to select what you see contextually.
> I can't think of a single real-world scenario when you'd want to select every item in the list while you've got a filter.

I agree that you don't want to "select all including those filtered out", but we should somehow express this in the UX (text?) when filtering is applied. Maybe something like "add filtered items to selection" (Huh, do we want to add the items to existing selection or do we want to replace the selection?).

### Filter-box - should it have a search icon and a clear icon? [IN PROGRESS]
> A search icon isn't required as it will update the list immediately as the user types.

This was meant as a visual indicator "this is the input for filtering" not as "spinner" while loading the data.
![image](https://github.com/havit/Havit.Blazor/assets/12828888/704e8b72-232d-4099-873c-1233291ee76a)

###  SelectAllChanged - do we even need such an EventCallback? How would it be used? If so, maybe OnSelectAllChanged, and it should not have a value.
> I have removed this.

Have you pushed all the changes? It is still there with all the related code.

### ItemSelectionChanged - we probably can't use this for the SelectAll scenario, as it would trigger a series of ValueChanged for each selected item, whereas it might be more appropriate to trigger ValueChanged just once with a List<TValue> of all items
>TBD

We probably have to change `SelectionChangedArgs` to allow multiple changes to be signaled at once. Probably something like
```csharp
public class SelectionChangedArgs
{
	public List<TItem> ItemsDeselected { get; set; }
	public List<TItem> ItemsSelected { get; set; }
}
```
The `HandleItemSelectionChanged` can then change the `CurrentValue` accordingly.

bholbrook commented:
@hakenr Haha, apparently I made the SelectAllChanged changes in my head! I've pushed them up now.

In terms of the SelectAll filtering language, I'm really not sure what to do here. If you have any suggestions I'm happy to make those changes.

I'm currently working on the Search/Clear icon. I didn't about having the Search icon there but that's not a bad idea.
I'll attempt to implement it similarly to the others whereby the Search icon will display while the filter box is empty and the Clear icon will display when there is any text.

And the ItemSelectionChanged stuff I haven't yet investigated. It's likely the last thing I'll work on.

bholbrook commented:
@hakenr I've just pushed up the final changes to your list of suggestions.
With them, everything should be ready to go.
If there is anything you think I should change (outside of documentation because I don't know how to run that project) please do let me know and I'll gladly make those changes.
I hope to get this integrated soon!

Also, once this does go in, how long might it be before this gets deployed to NuGet and made available?

bholbrook commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor:
```diff
@@ -36,25 +34,67 @@
-            for (int i = 0; i < ItemsToRender.Count; i++)
+            if (AllowFiltering)
+            {
+                <input @ref="filterInputReference"
```
I've wrapped the filter input and empty filter templates in `<li>` and have added a 0.25em padding in the filter input.

bholbrook commented on Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor:
```diff
@@ -36,25 +34,67 @@
+                <input @ref="filterInputReference"
+                       id="@($"{InputId}_filter")"
+                       type="text"
+                       class="form-control"
```
I have updated this such that the filter box now respects the parent input size.

hakenr reviewed (approved):
LGTM.
Will do some minor tuning and will merge to master.
